examples/hosts/app-integration/schema-upgrade/src/migrationInterfaces.ts

```diff
@@ -0,0 +1,89 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import type { IEvent, IEventProvider } from "@fluidframework/common-definitions";
+
+export interface IVersionedModel {
+    /**
+     * The string version of the model, matching the version of the container code it's paired with.
+     */
+    readonly version: string;
+}
+
+export interface IImportExportModel<ImportType, ExportType> {
+    /**
+     * Permit format checking in a generic manner - without knowing the type of our data or the type of the model,
+     * we can still check whether the model supports that data.
+     */
+    supportsDataFormat: (initialData: unknown) => initialData is ImportType;
+
+    /**
+     * importData must be called after initialization but before modifying or attaching the model (i.e. can only
+     * be called on an unaltered, detached model).
+     */
+    importData: (initialData: ImportType) => Promise<void>;
+
+    /**
+     * Export the data from the model.  Can be passed into importData() for a new container to replicate the data.
+     */
+    exportData: () => Promise<ExportType>;
+}
+
+export enum MigrationState {
+    collaborating,
+    migrating,
+    migrated,
+}
+
+export interface IMigratableModelEvents extends IEvent {
+    (event: "migrating" | "migrated", listener: () => void);
+}
+
+// TODO: Is there a better way to express the unknown format here?  I think I'd prefer to put the burden of calling
+// supportsDataFormat() on the callers of importData() (and allow implementers of IMigratableModel to assume
+// importData() is called with valid data).
+export interface IMigratableModel
+    extends IVersionedModel, IImportExportModel<unknown, unknown>, IEventProvider<IMigratableModelEvents> {
+    /**
+     * Get the current migration state of the model.
+     */
+    getMigrationState(): MigrationState;
+
+    /**
+     * The accepted migratory version, if migration has been accepted.
+     */
+    acceptedVersion: string | undefined;
+    /**
+     * Propose migration using the provided version.
+     * @param version - The version that the new model should use.
+     */
+    proposeVersion: (version: string) => void;
+
+    /**
+     * The containerId of the migrated model, if migration has completed.
+     */
+    newContainerId: string | undefined;
+    /**
+     * Complete the migration with the provided containerId.
+     * @param newContainerId - the ID of the container that the collaboration has migrated to.
+     */
+    finalizeMigration: (newContainerId: string) => Promise<void>;
+
+    /**
+     * Close the model, rendering it inoperable and closing connections.
+     * TODO: Decide whether the closing is an integral part of the migration, or if the caller should do the closing.
```

Comment:
It's probably true -- my main thought is that closing a container is a big, disruptive event that has a ton of side effects.  Depending on what an author is doing with the container, I could imagine it maybe being nice to have fine-grained control over when that happens (e.g. maybe they want to shut down some agent, or unregister some listeners, etc.).

What I'd really like to avoid is some `migrator.configureShouldCloseOnMigrationComplete(boolean)` -- basically if anyone might want the option, then I'd probably rather put it in their hands from the start.